### ISSUE TITLE
Shield Effects

### ISSUE THREAD

**Issue by gchapyzh** [endless-sky/endless-sky#2568]
I am sure this has been said before ( though I can't find it right now ) but is there a particular reason why there is no shield effects on ships?

Now, as I understand it, shields in Endless Sky are form fitting to the ship, and not just a bubble. There is a way to do that without making a separate shield texture for each ship - take a ship texture, turn it all to a monotone, see through color, and show a portion of shield when there is damage being dealt to it, at the vector that the damage is being dealt at. It may be difficult to implement, but it means that all future ships will come with shields automatically made.

Another object - different race's shield generators should give the shield different shades.

Are there reasons why this is not something we see yet?

**Comment by tehhowch** [comment]
Sounds great in practice, and could help folks who play without the overlay determine if a ship is taking hull damage or still only shield damage.

I imagine it may cause some flickering/render issues for a ship which is being hit by a great many particles from a great many directions at once, but the number of distinct vectors to draw the "ripple" could be limited in practice to around 20-30 total.

**Comment by gchapyzh** [comment]
Just divide the radius around the ship into sections a few degrees across, with neighboring once lighting up when one is triggered, and deepening in shade when more damage is applied.

**Comment by Amazinite** [comment]
I don't play with the overlay for a reason: it's distracting to me. I feel like this would be the same way. This would probably have to be toggleable in perferences.

**Comment by MessyMix** [comment]
I don't play with the overlay either. But I think this would be a cool effect, and I certainly would play with it! It should be in the preferences, though. :)

**Comment by gchapyzh** [comment]
As I see it, this would look a LOT more natural than any overlay could. As it is now, I feel weird seeing weapons hit my hull and damage shields instead, without any physical indication of said shields.

**Comment by jafdy** [comment]
It sounds great and we certainly have not had anyone bring an idea on how to do it before. My only concern is that this would require too much extra computing power. This would just make it a preference option.

**Comment by endless-sky** [comment]
This has been suggested before (#876) but the main difficulty would be finding a way to do it that wouldn't look tacky and also wouldn't use up too much computing power (because it would need to be a dynamically generated animation).

**Comment by KiLEdEnNis** [comment]
maybe adding an animated "shield layer" on every ship, although it'd look ugly if ya see like 20 ships with identical shield animation

**Comment by gchapyzh** [comment]
Well, the shield would not be visible unless damage is being applied. The game already refuses to overload on computing power with dozens of missile ships and hundreds of missiles in space with asteroids and things, I don't see how lighting several portions of a shield texture and stopping lasers and missiles there could be too much of an issue...

**Comment by PJTheGuy** [comment]
I would love to have this, it would make the game seem more realistic. It looks strange when a ship's shields are being damaged, but there isn't any visible effect from it.

**Comment by tehhowch** [comment]
If one peruses the issues/closed issue list, you'll note a lot of similar requests for shield indication. They're closed now, because no one volunteered their time to make it happen. MZ is plenty busy completing the story and code updates needed for the overall game direction, in addition to reviewing PRs others propose.

Anyone who chooses to develop shield indicators should 
1) Recognize that a merge-ready PR may still never get merged or commented on by MZ, or even be closed in favor of MZ's own implementation
2) Openly communicate development updates in their corresponding issue (with a clearly identified tasklist)
3) Implement the solution in such a manner that it meets or exceeds any criteria MZ may have expressed. For starters, in this issue all have been told it must not be a performance bottleneck, and it can't look tacky.


One possibility I can think of would be to perhaps define a `hit effect` for ships the way one does for weapons. This effect would then be rather small, colored in an appropriate manner lore-wise, and perhaps could be triggered in such a manner that it can be rotated to align with the ship's contour or some radial vector from the ships visual COM.  A close study of how things collide, how existing `hit effects` are generated and placed, the performance cost of existing hit effects, and so on - all of this is necessary. Don't forget that ship sprites can be animated as well, and your implementation will need to account for all of that variability as well.

**Comment by endless-sky** [comment]
There are simple mathematical models that simulate ripples traveling across the surface of a liquid. It might look cool to have each ship store an image that represents that sort of ripple simulation lattice, and whenever the ship is hit deform the lattice at the point where it was hit by an amount that reflects how much damage was done. (That's what I was thinking of when I said there could be a dynamically generated animation.) It would require one shader to step the image forward one frame, and another to mask out just the part of the lattice that's inside the ship and draw it on top in a suitable color.

**Comment by gchapyzh** [comment]
> @endless-sky - Techno-talk most of which I don't try to understand, but thanks for the explanation. 

So, as I see, its doable, one way or another. I think we can agree that this would be quite an improvement to the game, however, since right now it feels like shields don't exactly exist, and there is only hull/armor, judging purely by the looks of ships when they are attacked.

Different shield generators belonging to different races should also have distinct differences, so its clear and obvious that these are not the same shields ( even though mechanically they are ).

I recall seeing the words 'deflectors' somewhere in the descriptions, so it would be cool if maybe the beam and projectile weapons sometimes bounced at angles away from a shielded ship.

**Comment by MessyMix** [comment]
I don't think it would be a good idea to have deflection—in fact, none of the shields mention "deflect" in their description, and it would become really confusing for combat.

But, I agree that different kinds of shield generators should have visual differences. That would be awesome.

**Comment by jafdy** [comment]
What happens when you mix different generators? It might be better to have color based on the ship, shield matrices are built into the hull so that makes as much sense.

**Comment by MessyMix** [comment]
@jafdy that's a great point. I too would love a mixture of generators, but that might be complicated in the code.. by ship would be the next best option, and it makes logical sense too. :)

**Comment by KiLEdEnNis** [comment]
what if instead there's a basic version like nearly every ship in the game has the same explosions, but then you could make easy different shield effects, and you'd define which shield effect shows in the ships' stats?

**Comment by gchapyzh** [comment]
@MessyMix After some looking, I can indeed confirm that I must have imagined something - no mention of deflectors. I do wonder if that would be something the game could use too? A generator that gives beams and projectiles a certain percent chance to be deflected away at an angle?

@jafdy If you mix different generators, the effects should be mixed too - like mixing colors, different degrees of shield effect blurs, different ripple values, etc. Seems simple enough...

It already seems sketchy enough that different alien races all have compatible shield generators, though.

**Comment by Bladewood** [comment]
I want to be able to color the shield bar on displays purple and call it a `Barrier`. (From Mass Effect.)

**Comment by gchapyzh** [comment]
Deflector: Much lower strength value, slightly higher regen, but deflects (?) percent of inbound damage away from the ship.

**Comment by Nechochwen-D** [comment]
Not sure if anyone else has thought of this but what if we simply had a large texture that "flashes" into existence over the ship when it is hit with an alpha channel that is determined by the damage taken and current shield fraction. We could also give weapons a separate "shield impact" sprite for when they hit. Shield outfits could also have the option either override the default texture or add to the default texture.

**Comment by Bladewood** [comment]
Isn't that just OP's post?

**Comment by UnciaPrima** [comment]
Just wanted to +1 this. 
Adding some kind of visual clue  that shields are up and/or taking damage would be really cool

**Comment by Tadrix** [comment]
@jafdy ++ Also, there can be situation where ship has shields but no generator, thus ability to have shields is determined by the hull. But same here: "For the record I would prefer something based on the generator(s)."
